### PR TITLE
Implement Bits Per Byte Metric Computation

### LINKED ISSUE #60: [Feature] Bits per bytes metric
Implementation of the metric *bits per bytes* (BPB). BPB will allow us to compare both WELT and tokenizer-based methods in similar terms. This should be implement for both WELT's and tokenizer-based models' training scripts.

### PR DESCRIPTION
**Pull request by ilkerkesen**
## Fixes

Fixes #60  by @ilkerkesen 

## Description

Add bits-per-byte (BPB) metric to WeLT training and evaluation.

- Introduce `welt_training/metrics.py` with a `compute_bits_per_byte(loss, num_tokens, num_bytes)` utility
- Extend `WeLTTrainer` to accumulate per-batch nats and content byte counts during evaluation, computing BPB from exact token/byte counts (supports UTF-8, UTF-16, and UTF-32 encodings)
- Refactor accuracy computation from post-hoc batch-level reprocessing to incremental on-device accumulation during `prediction_step`, reducing memory overhead
- Add distributed evaluation support: `_sync_eval_state()` all-reduces scalar counters and gathers string predictions across ranks; handle last-batch padding via `_real_batch_count()`
- Add `_TorchIterableAdapter` to correctly wrap HF `IterableDataset` for PyTorch `DataLoader`
- Compute `eval_bits_per_byte` in `run_clm.py` for non-streaming CLM evaluation
- Report `eval_bits_per_byte` alongside existing `eval_loss`, `perplexity`, `eval_byte_accuracy`, and `eval_word_accuracy`.

## Tests

1. Unit tests for the BPB metric utility: `pytest tests/test_metrics.py -v`
2. Trainer integration tests (accuracy, BPB, distributed padding): `pytest tests/test_trainer.py -v`
3. Existing training tests (regression check): `pytest tests/test_train.py -v`
4. Run all tests together: `pytest tests/test_metrics.py tests/test_trainer.py tests/test_train.py -v`

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main` or `master`).
- [x] My commit messages follow the [contribution guidelines](../CONTRIBUTING.md).
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
      visible errors.

---

> [!NOTE]
> **Medium Risk**
> Touches `WeLTTrainer` evaluation internals (accuracy/BPB accumulation, distributed syncing, iterable dataloading), which could subtly change reported metrics or behavior under multi-GPU/streaming setups. Functionality is well-covered by new unit/integration tests, lowering regression risk.
> 
> **Overview**
> Adds a new `bits per byte` (BPB) evaluation metric via `compute_bits_per_byte()` and reports it as `eval_bits_per_byte` alongside loss/perplexity.
> 
> Refactors `WeLTTrainer` evaluation to incrementally accumulate byte/word accuracy and exact BPB numerators/denominators during `prediction_step` (including UTF-8/16/32 handling), then synchronizes counters and generated strings across distributed ranks before computing final metrics; also improves iterable/streaming eval dataloader handling (HF `IterableDataset` adapter + sharding) and fixes `eval_samples` counting.
> 
> Updates `run_clm.py` to compute `bits_per_byte` (and perplexity) during eval using per-token losses plus precomputed token/byte counts, and extends training/trainer tests to assert BPB presence/behavior and evaluation consistency.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 518d5bcce2c18dd5d05487481d4ccc717086c657. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added a bits-per-byte evaluation metric, byte- and word-level accuracy, and an option to report BPB alongside loss/perplexity.
  * New Gradio demo apps for interactive generation and entropy visualization; generation can optionally return per-byte entropy for plotting.
  * Evaluation: optional packed-eval mode and a preserve-document-boundaries option for chunked evaluation; improved iterable-dataset and distributed eval handling.

* **Refactor**
  * Updated internal type-check syntax.

* **Tests**
  * Extensive tests covering BPB math, tokenization/encoding effects, edge cases, and distributed consistency.

### CODE REVIEW

File: welt_training/trainer.py [hunk 1]
```diff
@@ -21,6 +21,22 @@
...
+    so PyTorch's DataLoader treats it as map-style and tries len()/__getitem__,
+    which fails. This thin adapter delegates __iter__ so DataLoader uses the
+    iterable protocol instead.
+    """
+
+    def __init__(self, hf_dataset):
+        self._dataset = hf_dataset
+
+    def __iter__(self):
+        yield from self._dataset
```
Comment (AmitMY) [inline welt_training/trainer.py:37]: i feel like this should be in `streaming.py`

**Review by AmitMY — CHANGES_REQUESTED** [review]
looks good! just one note regarding the streaming dataset

**Comment by AmitMY** [comment]
and tests fail

**Review by AmitMY — APPROVED** [review]

**Review by AmitMY — APPROVED** [review]

**Review by AmitMY — CHANGES_REQUESTED** [review]
lint and test fail
